### PR TITLE
feat!: Firebase iOS SDK version: `10.0.0`

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

1. nanopb dependency added to Firestore spec to fix zipped Firestore framework which was failing to build due to  undefined symbols from that library.
2. removed duplicate call of emulator as calling it twice causes a fatal error.
3. fixed `FIRStorageMetadata` bug causing a fatal error. Uploading file fails because it checks for [path](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/StorageUploadTask.swift#L67) property. `path` property is read only. But you can get around it by using:
```obj-c
[[FIRStorageMetadata alloc] initWithDictionary:metadata]
```
This allows you to set a `name` property which will be added to the metadata as a `path` property [here](https://github.com/firebase/firebase-ios-sdk/blob/970b4c45098319e40e6e5157d340d16cb73a2b88/FirebaseStorage/Sources/StorageMetadata.swift#L173-L175).

This needs to be flagged with iOS SDK team.

4. Updated podspec for plugins. iOS has a minimum deployment target of 11. macOS has a minimum deployment target of 10.13. [See documentation](https://firebase.google.com/support/release-notes/ios#version_1000_-_october_10_2022).

5. Skipped `getDownloadURL()` task as it fails when using the emulator. It fails [here](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/Internal/StorageGetDownloadURLTask.swift#L69-L70). I think because the serialised JSON response has a property that is not of the type `[String:String]`. Here is the error response:
```bash
value	String	"Invalid data returned from the server:{\"name\":\"flutter-tests/writeOnly.txt\",\"bucket\":\"flutterfire-e2e-tests.appspot.com\",\"generation\":\"1665653574035\",\"metageneration\":\"1\",\"contentType\":\"application/octet-stream\",\"timeCreated\":\"2022-10-13T09:32:54.035Z\",\"updated\":\"2022-10-13T09:32:54.035Z\",\"storageClass\":\"STANDARD\",\"size\":\"10\",\"md5Hash\":\"o8152dbQWHKzkt84nfUEKg==\",\"crc32c\":\"4120230975\",\"etag\":\"tHAEijHjZXFls/6B1nPFt/DJ7Ec\",\"downloadTokens\":\"97fec822-84bd-422f-a6ab-82dc30de6029\",\"contentEncoding\":\"identity\",\"contentDisposition\":\"inline\",\"metadata\":{}}"	
``` 
Note that `metadata` has an empty object as its value. When making the same request on the Firebase server (not the emulator), we get a successful response.  It does not contain the property `metadata`. See:
Screenshot 2022-10-13 at 10 37 15

I have skipped this test for now but it is something to flag with iOS SDK team.

6. Cannot `initWithDictionary()` when setting metadata (`updateMetadata()` ) which is a bug. `initWithDictionary()` will set `initialMetadata` [here](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/StorageMetadata.swift#L157). When `StorageUpdateMetadataTask` calls `updateMetadata.updatedMetadata()` [here](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/Internal/StorageUpdateMetadataTask.swift#L54), it will compare `initialMetadata` and the returned dictionary of `dictionaryRepresentation()` [here](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/StorageMetadata.swift#L112-L128). It then calls a `remove()` function that compares the values and set the values to `nil` if "old" metadata property and "new" metadata property are the same. See:
Screenshot 2022-10-13 at 11 34 11

I don't understand why it sets the values to `nil`. I also don't understand how it can be "new" & "old" metadata when it's the same metadata used to initialize `FIRStorageMetadata`. Either way, we can get around this problem by simply setting `customMetadata` property the way we have done previously:

```obj-c
FIRStorageMetadata *metadata = [[FIRStorageMetadata alloc] init];
metadata.customMetadata = dictionary[kFLTFirebaseStorageKeyCustomMetadata];
```
This should also be flagged with iOS SDK team.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database](https://github.com/flutter/flutter/issues). Indicate, which of these issues are resolved or fixed by this PR. Note that you'll have to prefix the issue numbers with flutter/flutter#.*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m [hunk 1]
```diff
@@ -191,7 +193,24 @@ - (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)flutter
...
+  NSString *appName = arguments[kFLTFirebaseStorageKeyAppName];
+  NSString *bucket = arguments[kFLTFirebaseStorageKeyBucket];
+  FIRApp *firebaseApp = [FLTFirebasePlugin firebaseAppNamed:appName];
+
+  if (![bucket isEqual:[NSNull null]] && bucket != nil) {
+    NSString *url = [@"gs://" stringByAppendingString:bucket];
+    storage = [FIRStorage storageForApp:firebaseApp URL:url];
+  } else {
+    storage = [FIRStorage storageForApp:firebaseApp];
+  }
```
Comment (russellwheatley) [inline packages/firebase_storage/firebase_storage/ios/Classes/FLTFirebaseStoragePlugin.m:206]: Not sure if all this config is needed 🤔.

File: packages/firebase_core/firebase_core/ios/firebase_core.podspec [hunk 2]
```diff
@@ -27,7 +27,7 @@ Pod::Spec.new do |s|
   s.source_files     = 'Classes/**/*'
   s.public_header_files = 'Classes/**/*.h'
 
-  s.ios.deployment_target = '9.0'
+  s.ios.deployment_target = '11.0'
```
Comment (Lyokone) [inline packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: I see on https://github.com/firebase/firebase-ios-sdk/blob/174be0d92bd6706b30a7a9d18f6bbe2319809e03/Firebase.podspec that the ios.deployment_target is still 10.0.

Do we need to manually add the deployment target? Or should CocoaPod just pick it from the dependency? It would me one less thing to maintain from our side

Comment (russellwheatley) [reply packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: I see what you mean, but see what @Salakar thinks as I'm not sure.

Comment (Salakar) [reply packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: I believe we can remove it or at a minimum set it back to 10.0

Comment (russellwheatley) [reply packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: Removing deployment targets caused [build failures](https://github.com/firebase/flutterfire/actions/runs/3264629221) so I guess we do need them. Not sure we should go back to version 10, @Salakar.  [See documentation](https://firebase.google.com/support/release-notes/ios#version_1000_-_october_10_2022). Also - look at the CI logs. Sample:

```
of supported deployment target versions is 11.0 to 16.0.99. (in target 'FirebaseAppCheckInterop' from project 'Pods')
    /Users/runner/work/flutterfire/flutterfire/tests/ios/Pods/Pods.xcodeproj: warning: The iOS Simulator deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 10.0, but the range of supported deployment target versions is 11.0 to 16.0.99. (in target 'GoogleAppMeasurement' from project 'Pods')
```

Comment (Salakar) [reply packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: The pod relies on 10.0 and above still, but if the docs say they want to support 11+ then sticking to 11 here is fine by me!

Comment (russellwheatley) [reply packages/firebase_core/firebase_core/ios/firebase_core.podspec:30]: I guess this makes it a breaking change now across all plugins. I'm also a little concerned about the behaviour of Storage (it's completely changes to a Swift implementation) as I wonder what other edge cases have changed (e.g. exceptions/errors) that we aren't capturing in our CI.

File: packages/firebase_storage/firebase_storage/example/ios/Podfile [hunk 3]
```diff
@@ -28,6 +28,8 @@ require File.expand_path(File.join('packages', 'flutter_tools', 'bin', 'podhelpe
 flutter_ios_podfile_setup
 
 target 'Runner' do
+  use_frameworks!
+  use_modular_headers!
```
Comment (mikehardy) [inline packages/firebase_storage/firebase_storage/example/ios/Podfile:32]: I know use_frameworks! is a strict requirement of firebase-ios-sdk these days, but I'm unfamiliar with `use_modular_headers` being required. Is modular headers required now as well?

File: tests/pubspec.yaml [hunk 4]
```diff
@@ -8,12 +8,6 @@ environment:
   sdk: ">=2.16.0 <3.0.0"
 
 dependencies:
-  cloud_firestore:
-    path: ../packages/cloud_firestore/cloud_firestore
-  cloud_firestore_platform_interface:
-    path: ../packages/cloud_firestore/cloud_firestore_platform_interface
-  cloud_firestore_web:
-    path: ../packages/cloud_firestore/cloud_firestore_web
```
Comment (mikehardy) [inline tests/pubspec.yaml:16]: is it intended + desired to remove all these from the pubspec? I'm not sure exactly what's going on or why (apologies) I just am not used to seeing a negative line count in "tests" areas as a positive :thinking:

File: tests/test_driver/firebase_storage/reference_e2e.dart [hunk 5]
```diff
@@ -127,17 +127,23 @@ void setupReferenceTests() {
...
+          await ref.putString('ok');
+
+          String downloadUrl = await ref.getDownloadURL();
+          expect(downloadUrl, isA<String>());
+          expect(downloadUrl, contains('ok.txt'));
+          expect(downloadUrl, contains(storage.app.options.projectId));
+        },
+        // Fails on emulator since iOS SDK 10. See PR notes:
+        // https://github.com/firebase/flutterfire/pull/9708
+        skip: defaultTargetPlatform == TargetPlatform.iOS ||
```
Comment (mikehardy) [inline tests/test_driver/firebase_storage/reference_e2e.dart:143]: Ah - read the notes, also about the `path` thing. That's a shame! I worked really really hard with Paul and the emulator folks a few months back and we had the storage metadata stuff locked down after a couple rounds of "probe until we understand" + "fix either react-native-firebase or firebase-ios-sdk or firebase-tools". There were a few little corner cases on all sides. May take a bit to tack down the regressions, I can help here as well, as react-native-firebase has a very thorough StorageMetadata test setup after all that

File: tests/pubspec.yaml (hunk 4 above)
Comment (Salakar) [reply tests/pubspec.yaml:16]: We test cloud firestore in a separate app & CI job, since it's a biggie

File: tests/test_driver/firebase_storage/reference_e2e.dart (hunk 5 above)
Comment (Salakar) [reply tests/test_driver/firebase_storage/reference_e2e.dart:143]: one other thing to note is since storage has moved to swift and they now use `fatalError` we can no longer catch things like 'emulator already set' etc, as more things move to swift some of our try/catchers for workarounds are going to be more problematic

File: tests/ios/Podfile [hunk 6]
```diff
@@ -46,8 +46,6 @@ target 'Runner' do
       raise "Error - unable to locate firebase_ios_sdk.rb script in firebase_core, and no FirebaseSDKVersion specified"
     end
   end
-
-  pod 'FirebaseFirestore', :git => 'https://github.com/invertase/firestore-ios-sdk-frameworks.git', :tag => "#{firebase_sdk_version}"
```
[the remark this answers is not in the document]
Comment (russellwheatley) [reply tests/ios/Podfile:50]: Interesting, yeah, maybe it should be on the framework actually. If you do update the framework, could you let me know and I can experiment by taking it away as a dependency on FirebaseFirestore? 😄 I imagine RNFB will have the same issue 🤔

File: tests/pubspec.yaml (hunk 4 above)
Comment (russellwheatley) [reply tests/pubspec.yaml:16]: Firestore has its own test suite in its directory. I believe this was never removed when we ported over to a separate test app for all the other  plugins (similar to test setup for RNFB) which increases build time.

File: packages/firebase_storage/firebase_storage/example/ios/Podfile (hunk 3 above)
Comment (russellwheatley) [reply packages/firebase_storage/firebase_storage/example/ios/Podfile:32]: I've just tested without it and I guess it isn't needed. I was under the impression it was needed along with `use_frameworks!` 👍

File: tests/pubspec.yaml (hunk 4 above)
Comment (russellwheatley) [reply tests/pubspec.yaml:16]: (didn't see Mike's comment 🙈 )

File: tests/test_driver/firebase_storage/reference_e2e.dart (hunk 5 above)
Comment (mikehardy) [reply tests/test_driver/firebase_storage/reference_e2e.dart:143]: Please tag me in that @russellwheatley - I was neck-deep in it so recently I can hopefully assist

File: tests/ios/Podfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikehardy) [reply tests/ios/Podfile:50]: roger - will do

**Comment by russellwheatley** [comment]
> 🤔
> 
> ```
> 
>   firebase_dynamic_links buildShortLink
>      ✘ build short dynamic links
>        [E]: type 'int' is not a subtype of type 'String' in type cast
> 
>   firebase_dynamic_links getInitialLink
>      ✔ initial link
> 
>   firebase_dynamic_links getDynamicLink
>      ✘ dynamic link using uri created on Firebase console
>        [E]: type '_InternalLinkedHashMap<Object?, Object?>' is not a subtype of type 'String' in type cast
>      ✔ Universal link error for URL that cannot be parsed
> ```
> 
> [firebase/flutterfire/actions/runs/3265837960/jobs/5369204292#step:12:1040](https://github.com/firebase/flutterfire/actions/runs/3265837960/jobs/5369204292#step:12:1040)

@mikehardy yeah, they've just recently started to fail 😓. I'm not sure why. I'm guessing something to do with Firebase Server. Investigating now.

**Comment by russellwheatley** [comment]
Whitelisting deeplink domain on Firebase console has solved one of the tests. I think it'll be something along those lines for the other test as well.

File: packages/cloud_firestore/cloud_firestore/ios/cloud_firestore.podspec [hunk 7]
```diff
@@ -24,17 +24,18 @@ Pod::Spec.new do |s|
...
   s.private_header_files = 'Classes/Private/*.h'
 
-  s.ios.deployment_target = '10.0'
+  s.ios.deployment_target = '11.0'
   s.dependency 'Flutter'
 
   s.dependency 'firebase_core'
   s.dependency 'Firebase/Firestore', firebase_sdk_version
-  
+  s.dependency 'nanopb'
```
Comment (mikehardy) [inline packages/cloud_firestore/cloud_firestore/ios/cloud_firestore.podspec:37]: I logged an issue with the appropriate details, and the dependency here should have the same version constraint as upstream I think?

```suggestion
  # required until firestore-ios-sdk-frameworks is updated, otherwise users of that distribution will have compile failures
  # see https://github.com/invertase/firestore-ios-sdk-frameworks/issues/59
  s.dependency 'nanopb', '>= 2.30908.0', '< 2.30910.0'
```

File: packages/cloud_firestore/cloud_firestore/macos/cloud_firestore.podspec [hunk 8]
```diff
@@ -56,6 +56,7 @@ Pod::Spec.new do |s|
   s.dependency 'firebase_core'
   s.dependency 'Firebase/CoreOnly', "~> #{firebase_sdk_version}"
   s.dependency 'Firebase/Firestore', "~> #{firebase_sdk_version}"
+  s.dependency 'nanopb'
```
Comment (russellwheatley) [inline packages/cloud_firestore/cloud_firestore/macos/cloud_firestore.podspec:59]:
```suggestion
  # required until firestore-ios-sdk-frameworks is updated, otherwise users of that distribution will have compile failures
  # see https://github.com/invertase/firestore-ios-sdk-frameworks/issues/59
  s.dependency 'nanopb', '>= 2.30908.0', '< 2.30910.0'
```

**Review by pr-Mais — APPROVED** [review]

File: tests/test_driver/firebase_storage/reference_e2e.dart (hunk 5 above)
Comment (mikehardy) [inline tests/test_driver/firebase_storage/reference_e2e.dart:143]: For tracking - this will resolve with https://github.com/firebase/firebase-ios-sdk/pull/10370 (or perhaps something in firebase-tools)

**Comment by JavierPerezLavadie** [comment]
Hello, I don't understand, you approved a commit with a storage bug in the emulator ?
Now i can't use the emulator for my app anymore, do you have an alternative to getDownloadURL() on the emulator?

Skipped getDownloadURL() task as it fails when using the emulator. It fails [here](https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorage/Sources/Internal/StorageGetDownloadURLTask.swift#L69-L70). I think because the serialised JSON response has a property that is not of the type [String:String]. Here is the error response:

**Comment by mikehardy** [comment]
@JavierPerezLavadie This is a very large project, and the storage emulator release schedule is not in this project's control. On balance, sometimes software ships with known issues and for those affected they need to handle with extra care.

I believe you can use the storage emulator *by not upgrading if the issue affects you*. Does it work when you stay on the versions of flutter fire prior to the release?

**Comment by JavierPerezLavadie** [comment]
@mikehardy 

I understand but this problem makes the storage on the emulator unusable.

I should have waited a bit longer before updating.

Yes before the update it worked.

I saw here that you already fixed the error?

https://github.com/firebase/firebase-ios-sdk/issues/10374

Do you think these PRs for Putfile and getDownload() for the emulator will be in the next update?

Thanks

**Comment by mikehardy** [comment]
> @mikehardy 
> 
> I understand but this problem makes the storage on the emulator unusable.

That is use case dependent, I understand most will probably use the affected method but not all. Not to quibble just to quibble but beware blanket statements in software

On the plus side firebase has counts functionality now which is a big deal as announced at firebase summit. You can imagine the desire for some to start incorporating that and you can see the competing pressure for releases

> 
> I should have waited a bit longer before updating.

Source control hopefully makes the revert easy after qualification on the new release fails. It is hassle, no doubt, apologies for that.


> 
> Yes before the update it worked.
> 
> I saw here that you already fixed the error?
> 
> https://github.com/firebase/firebase-ios-sdk/issues/10374
> 
> Do you think these PRs for Putfile and getDownload() for the emulator will be in the next update?



They are both merged so to the best of my know they will be in firebase-ios-sdk release 10.1.0 yes (or 10.0.1, next release basically)


I've never tried integration with the upstream nightly or beta releases but perhaps it is possible to access those by overriding something in firebase core as well https://github.com/firebase/firebase-ios-sdk/tags depending on the urgency for you

**Comment by JavierPerezLavadie** [comment]
Thank you very much for your detailed answer, I will wait for the next update hoping that it will solve the problem.
